core/src/main/java/org/apache/calcite/sql/SqlSelect.java

```diff
@@ -288,4 +297,133 @@ public boolean hasWhere() {
   public boolean isKeywordPresent(SqlSelectKeyword targetKeyWord) {
     return getModifierNode(targetKeyWord) != null;
   }
+
+  /**
+   * Rewrite a select statement with a where condition if the where
+   * condition depends on the columns in the select list. This writes
+   * the code to use two select statements.
+   *
+   * For example <code>Select A + 1 as X from table where X > 3</code>
+   * becomes
+   * <code>Select TEMP_COLUMN0 FROM (Select A + 1 as TEMP_COLUMN0)</code>
+   * <code>where TEMP_COLUMN0 > 3</code>
+   *
+   * We only do this replacement if we detect that the where may reference an alias generated
+   * by the select statement. This approach simplifies code but is not required for correctness.
+   * If this fails then the alias will be inlined as a fallback in validation.
+   *
+   * @return The new top level select
+   */
+  public SqlSelect rewriteSqlSelectIfWhereAlias() {
```

Comment:
I think there's likely an edgecase here. Assume table1 has columns A, B, C

`SELECT expr as A from table1 where A > 3`

NOTE: In this case, the where clause should now be `where A > 3`

This will be rewritten into:

`Select A from (Select expr as A, A, B, C from table1) where A > 3`

I think that the outer query will fail to validate, since "A" is now ambiguous in the outer query, since the inner query has two columns named "A".

I think this can be resolved by returning in the case that we see a possible alias that matches one of the input's rows (see below).

Reply:
That sounds good. I can keep two sets, known aliases and known identifiers and if an alias matches an identifier and an alias we cannot handle it.